<a href="https://colab.research.google.com/github/Ishikaaa/ML-Data-Scientist-job-market-in-the-US/blob/master/Data_Scientist_job_market_in_the_US.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ***1. Importing Libraries - ***

In [0]:
import pandas as pd 
import numpy as np 
import re 
import seaborn as sns 

In [53]:
dataset=pd.read_csv("alldata.csv")
dataset.head()

position  ...            location
0                               Development Director  ...  Atlanta, GA 30301 
1  An Ostentatiously-Excitable Principal Research...  ...         Atlanta, GA
2                                     Data Scientist  ...         Atlanta, GA
3                                       Data Analyst  ...  Atlanta, GA 30303 
4  Assistant Professor -TT - Signal Processing & ...  ...         Atlanta, GA

[5 rows x 5 columns]

In [54]:
dataset.shape 

(6964, 5)

In [55]:
dataset.dtypes

position        object
company         object
description     object
reviews        float64
location        object
dtype: object

# ***Handling Missing Data - ***

In [56]:
dataset.isnull().sum()

position         11
company          11
description      11
reviews        1638
location         11
dtype: int64

In [57]:
# removing that rows in which all colunms are null 
dataset = dataset.dropna(how='all',axis=0)
dataset.shape

(6953, 5)

In [58]:
dataset.isnull().sum()

position          0
company           0
description       0
reviews        1627
location          0
dtype: int64

In [59]:
# remove that rows which do not have any review 
dataset=dataset.fillna(0)
dataset.shape 

(6953, 5)

In [60]:
dataset.isnull().sum()

position       0
company        0
description    0
reviews        0
location       0
dtype: int64

In [61]:
dataset['position'].value_counts()

Data Scientist                                                                                                              351
Senior Data Scientist                                                                                                        96
Research Analyst                                                                                                             64
Data Engineer                                                                                                                60
Machine Learning Engineer                                                                                                    56
Lead Data Scientist                                                                                                          31
Research Scientist                                                                                                           30
Sr. Data Scientist                                                                                      

In [62]:
dataset['company'].value_counts()

Amazon.com                                358
Ball Aerospace                            187
Microsoft                                 137
Google                                    134
NYU Langone Health                         77
Fred Hutchinson Cancer Research Center     70
KPMG                                       66
Lab126                                     50
Facebook                                   49
Broad Institute                            49
Walmart eCommerce                          45
McKinsey & Company                         41
Takeda Pharmaceutical                      39
University of Washington                   39
Harnham                                    39
Oath Inc                                   34
Pharmacyclics, an Abbvie Company           34
Celgene Corporation                        34
Cymer                                      32
Brigham & Women's Hospital(BWH)            31
Biogen                                     30
Amgen                             

In [63]:
dataset['reviews'].value_counts()

0.0        1627
25902.0     224
108.0       193
2.0         191
5.0         138
4.0         131
4618.0      118
2474.0      106
3.0         101
6.0          94
25895.0      89
7.0          78
705.0        76
91.0         71
8.0          66
15.0         65
838.0        62
32.0         55
11.0         54
366.0        49
14.0         48
141.0        46
12.0         45
26.0         42
70.0         42
10.0         41
385.0        41
52.0         39
530.0        39
17.0         36
           ... 
99.0          1
4338.0        1
242.0         1
2203.0        1
25909.0       1
2189.0        1
415.0         1
694.0         1
723.0         1
1516.0        1
312.0         1
55.0          1
49.0          1
3027.0        1
11607.0       1
1947.0        1
131.0         1
177.0         1
63.0          1
228.0         1
892.0         1
2431.0        1
2647.0        1
447.0         1
1959.0        1
1366.0        1
6868.0        1
21949.0       1
290.0         1
2409.0        1
Name: reviews, Length: 5

In [64]:
dataset['description'][0]

"Development Director\nALS Therapy Development Institute has an immediate opening for Development Directors. Reporting directly to the Senior Development Director, the Development Director at ALS TDI is a senior fundraising position working to identifying potential prospects and cultivating solicitation strategies and in closing asks with donors including individuals and corporations by building networks via events, generating awareness of ALS TDI; outreach including attending and speaking at events as well as personally cultivates relationships with patients, prospects and donors. This position will be responsible for generating and managing a portfolio of at least two million to five million dollars per year. This position will be located in Atlanta, GA.\n\nRequirements:\nBachelor's Degree requiredMinimum 6-8 years' experience in fundraising and business developmentSuccessful track recording in fundraising with major donors or scientific sales preferredDemonstrated ability to work in

In [68]:
dataset['location'].value_counts()

Seattle, WA                563
New York, NY               508
Cambridge, MA              487
Boston, MA                 454
San Francisco, CA          425
San Diego, CA              294
Chicago, IL                292
Boulder, CO                233
Washington, DC             230
Sunnyvale, CA              221
Mountain View, CA          199
Los Angeles, CA            187
Austin, TX                 156
Atlanta, GA                147
Redmond, WA 98052          117
Cambridge, MA 02139         95
Seattle, WA 98109           95
New York, NY 10016          83
San Diego, CA 92121         71
Cambridge, MA 02142         64
South San Francisco, CA     53
Chicago, IL 60654           45
San Francisco, CA 94103     40
Chicago, IL 60601           40
Mountain View, CA 94043     39
Redmond, WA                 39
Seattle, WA 98101           34
Seattle, WA 98104           34
Cambridge, MA 02140         33
Boston, MA 02116            31
                          ... 
Jersey City, NJ 07310        1
Atlanta,

***Note - ***

1.   Most of the job offers are on the location 'Seattle, WA'.



In [69]:
dataset[dataset['reviews']==dataset['reviews'].max()]

position  ...             location
6728  Staff Software Engineer, Front End Lead  ...  Sunnyvale, CA 94086
6828         Information Security Risk Expert  ...  Sunnyvale, CA 94086
6958         Senior Data Scientist (GEC11902)  ...  Sunnyvale, CA 94086

[3 rows x 5 columns]

In [70]:
dataset[dataset['company']=='Walmart']

position  ...             location
449                        Data Scientist-ISD  ...     Austin, TX 78716
6668                     Staff Data Scientist  ...  San Bruno, CA 94066
6728  Staff Software Engineer, Front End Lead  ...  Sunnyvale, CA 94086
6828         Information Security Risk Expert  ...  Sunnyvale, CA 94086
6958         Senior Data Scientist (GEC11902)  ...  Sunnyvale, CA 94086

[5 rows x 5 columns]

***Note - ***

1.   Company "Walmart" has maximum reviews. And Company "Walmart" offers maximum jobs on location "Sunnyvale, CA 94086".

In [71]:
dataset[dataset['reviews']==dataset['reviews'].min()]

position  ...             location
0                                  Development Director  ...   Atlanta, GA 30301 
1     An Ostentatiously-Excitable Principal Research...  ...          Atlanta, GA
2                                        Data Scientist  ...          Atlanta, GA
18                                       Data Scientist  ...          Atlanta, GA
19                                       Data Scientist  ...          Atlanta, GA
25                                       Data Scientist  ...          Atlanta, GA
26                 Data Scientist 1 – AP Forensics Team  ...          Atlanta, GA
38                                       Data Scientist  ...          Atlanta, GA
48                        Chief Deep Learning Scientist  ...          Atlanta, GA
53                                  Statistical Analyst  ...          Atlanta, GA
54                                   VP, Data Analytics  ...          Atlanta, GA
55                              Lead Decision Scientist  ...   Atlanta, GA 30313 
65                                   Research Scientist  ...          Atlanta, GA
84                                   Research Scientist  ...          Atlanta, GA
106   Staff Research Scientist VII: Informational In...  ...          Atlanta, GA
114                             Jr Microbiologist - CDC  ...          Atlanta, GA
122               Mid Data Scientist - Machine Learning  ...   Atlanta, GA 30342 
139   Principal Research Scientist I: Informational ...  ...          Atlanta, GA
140                             Market Research Analyst  ...          Atlanta, GA
151                        Senior Scientist, Immunology  ...          Atlanta, GA
153          A Vigorously-Charged Scientific Programmer  ...          Atlanta, GA
154                                   Software Engineer  ...          Atlanta, GA
156                                Director of Software  ...   Atlanta, GA 30332 
157       Certifying Scientist / Quantitation Scientist  ...          Atlanta, GA
161                                      Data Scientist  ...   Atlanta, GA 30326 
164   Windows Programmer / Software Engineer - Inter...  ...          Atlanta, GA
170   Windows Programmer / Software Engineer - Entry...  ...          Atlanta, GA
192   Various Professional, Technical, Administrativ...  ...          Atlanta, GA
196     Business Development Manager (Southeast Region)  ...          Atlanta, GA
199   Implementation Engineer (Machine Learning/Data...  ...   Atlanta, GA 30326 
...                                                 ...  ...                  ...
6789                                     Head of Design  ...  Sunnyvale, CA 94085
6791                 Forward Deployed Solution Engineer  ...        Sunnyvale, CA
6795        Deep Learning Software Development Engineer  ...        Sunnyvale, CA
6800                                 Knowledge Engineer  ...        Sunnyvale, CA
6813  AI Research/Software Development Engineer, Ale...  ...        Sunnyvale, CA
6815                         3D Tools Software Engineer  ...  Sunnyvale, CA 94089
6816       Software Engineer - Tools and Infrastructure  ...        Sunnyvale, CA
6817  23103 Principal Data Scientist - HealthTech / ...  ...        Sunnyvale, CA
6822                                      Manager, TPMs  ...        Sunnyvale, CA
6824                               QA Software Engineer  ...        Sunnyvale, CA
6825                                Sr Speech Scientist  ...        Sunnyvale, CA
6827                      Software Development Engineer  ...        Sunnyvale, CA
6829                                  Software Engineer  ...        Sunnyvale, CA
6841                      Sr. Machine Learning Engineer  ...        Sunnyvale, CA
6846                        Sr. Hardware Audio Engineer  ...        Sunnyvale, CA
6848                       Materials (Quality) Engineer  ...  Sunnyvale, CA 94086
6856  Senior Data Scientist, Algorithms and Machine ...  ...        Sunnyvale, CA
6868                    Senio

In [72]:
dataset.head()

position  ...            location
0                               Development Director  ...  Atlanta, GA 30301 
1  An Ostentatiously-Excitable Principal Research...  ...         Atlanta, GA
2                                     Data Scientist  ...         Atlanta, GA
3                                       Data Analyst  ...  Atlanta, GA 30303 
4  Assistant Professor -TT - Signal Processing & ...  ...         Atlanta, GA

[5 rows x 5 columns]

# ***Adding new columns - ***

In [74]:
# Adding new column "City"
a=[]
for i in dataset["location"]:
    b=i.split(",")[0]
    a.append(b)
dataset["City"]=a    
dataset.tail()

position  ...       City
6959  Data Developer / Machine Learning Analyst  ...  Sunnyvale
6960                                Scientist I  ...  Sunnyvale
6961                           Intern Scientist  ...  Sunnyvale
6962            Senior Data & Applied Scientist  ...  Sunnyvale
6963    Principal Data Scientist, Deep Learning  ...  Sunnyvale

[5 rows x 6 columns]

In [82]:
dataset["City"].value_counts()

New York                  848
Seattle                   777
Cambridge                 694
Boston                    629
San Francisco             564
Chicago                   471
San Diego                 412
Washington                340
Mountain View             277
Atlanta                   269
Sunnyvale                 256
Los Angeles               254
Boulder                   251
Austin                    213
Redmond                   158
South San Francisco        73
Redwood City               59
Oakland                    35
San Mateo                  30
Manhattan                  30
San Francisco Bay Area     26
Emeryville                 19
Berkeley                   19
Jersey City                17
San Carlos                 16
Queens                     15
Brooklyn                   14
Menlo Park                 14
Foster City                14
Novato                     12
                         ... 
New Brunswick               1
Long Beach                  1
Cedar Park

In [80]:
# Adding new column "State"
aa=[]
for ii in dataset["location"]:
    bb=ii.split(",")[1]
    aa.append(bb)
for ii in range(len(aa)):
    bb=re.sub(r'[0-9]',"",aa[ii]).strip()
    aa[ii]=bb  
dataset["State"]=aa
dataset.tail()

position  ... State
6959  Data Developer / Machine Learning Analyst  ...    CA
6960                                Scientist I  ...    CA
6961                           Intern Scientist  ...    CA
6962            Senior Data & Applied Scientist  ...    CA
6963    Principal Data Scientist, Deep Learning  ...    CA

[5 rows x 7 columns]

In [81]:
dataset["State"].value_counts()

CA    2152
MA    1323
WA     935
NY     926
IL     471
DC     340
GA     269
CO     251
TX     218
NJ      68
Name: State, dtype: int64

In [83]:
# adding a new column "Data Scientists"
aaa=[]
for i in dataset["position"]:
    aaa.append(("Data Scientist" in i))
dataset["Data Scientist"]=aaa
dataset.head()

position  ... Data Scientist
0                               Development Director  ...          False
1  An Ostentatiously-Excitable Principal Research...  ...          False
2                                     Data Scientist  ...           True
3                                       Data Analyst  ...          False
4  Assistant Professor -TT - Signal Processing & ...  ...          False

[5 rows x 8 columns]

In [84]:
dataset["Data Scientist"].value_counts()

False    5710
True     1243
Name: Data Scientist, dtype: int64

# ***Analysis From Data - ***

***Analysis upto now - ***


1.   Most of the job offers are on the location 'Seattle, WA'.
2.   Company "Walmart" has maximum reviews. And Company "Walmart" offers maximum jobs on location "Sunnyvale, CA 94086".



In [0]:
# which position most "Walmart" company offers. 
# which company offers maximum "Data Scientist" job .
# which positions are offeres at location "Seattle,WA".

In [87]:
dataset[dataset["company"]=="Walmart"].value_counts()

AttributeError: ignored

In [107]:
print(dataset["description"][3])

DEPARTMENT: Program OperationsPOSITION LOCATIONS: Atlanta, GA.ACCOUNTABILITY:  President of Program OperationsFor 25 years Operation HOPE has been dedicated to empowering underserved communities through financial literacy, now our work encompasses more than financial education. Our work is about instilling knowledge and confidence in those we serve so they can experience, sometimes for the first time in their lives, financial dignity.JOB SUMMARYThe data analyst collects, organizes, mines, analyzes, and audits large sets of data for impact reporting and operational optimization. The Data Analyst participates as a team member responsible for maintaining the integrity of client database information and provides executable and actionable recommendations for data migration best practices and field operations based on analytical findings.DUTIES AND RESPONSIBILITIESInitiate and participate in data mining and reportingAudit and profile the data to assess the impact of poor quality data on the 